Docs update (#73)
solidity-labs-io/forge-proposal-simulator: docs/guides/multisig-proposal.md

## Overview

After adding FPS into project dependencies, the next step is the creation of the
first Proposal contract. This example provides guidance on writing a proposal
for deploying new instances of `Vault.sol` and `Token`. These contracts are

> located in the fps-example-repo [mocks](https://github.com/solidity-labs-io/fps-example-repo/tree/main/src/mocks). Copy each file used in this tutorial into your project for running examples or clone [fps-example-repo](https://github.com/solidity-labs-io/fps-example-repo/) repo and you have everything setup there. For this tutorial it's assumed that you have cloned the fps-example-repo.

Reviewer: same comments as governor bravo proposal

Author: fixed

This proposal includes the transfer of ownership of both contracts to multisig, along with the whitelisting of the token, minting of tokens to the multisig and multisig depositing tokens into the vault.

## Proposal contract

> The following contract is present in the [src/proposals/](https://github.com/solidity-labs-io/fps-example-repo/tree/main/src/proposals) folder. We will use this contract as a reference for the tutorial.

Reviewer: link to the file instead of to the folder

Author: done

            multisigVault.transferOwnership(multisig);
        }

        if (!addresses.isAddressSet("MULTISIG_TOKEN")) {
            Token token = new Token();
            addresses.addAddress("MULTISIG_TOKEN", address(token), true);
            token.transferOwnership(multisig);

            // During forge script execution, the deployer of the contracts is
            // the DEPLOYER_EOA. However, when running through forge test, the deployer of the contracts is this contract.
            uint256 balance = token.balanceOf(address(this)) > 0
                ? token.balanceOf(address(this))
                : token.balanceOf(addresses.getAddress("DEPLOYER_EOA"));

            token.transfer(multisig, balance);
        }
    }

    function build()
        public
        override
        buildModifier(addresses.getAddress("DEV_MULTISIG"))
    {
        address multisig = addresses.getAddress("DEV_MULTISIG");

        /// STATICCALL -- not recorded for the run stage
        address multisigVault = addresses.getAddress("MULTISIG_VAULT");
        address token = addresses.getAddress("MULTISIG_TOKEN");
        uint256 balance = Token(token).balanceOf(address(multisig));

>         Vault(multisigVault).whitelistToken(token, true);

Reviewer: this should be under the

`        /// CALLS -- mutative and recorded`

section

Author: done

## Overview

> After adding FPS into project dependencies, the next step is the creation of the first Proposal contract. This example provides guidance on writing a proposal for deploying new instances of `Vault.sol` and `Token`. These contracts are located in the fps-example-repo [mocks](https://github.com/solidity-labs-io/fps-example-repo/tree/main/src/mocks). Copy each file in this tutorial into your project, or clone the [fps-example-repo](https://github.com/solidity-labs-io/fps-example-repo/) repo. This tutorial assumes you have cloned the fps-example-repo.

Reviewer: the example does more than just deploy the contracts, it also performs actions to set them up via trusted actors through after deploy actions and the build and run function

Author: Done

> After adding FPS into project dependencies, the next step is the creation of the first Proposal contract. This example provides guidance on writing a proposal for deploying new instances of `Vault.sol` and `Token`. These contracts are located in the fps-example-repo [mocks](https://github.com/solidity-labs-io/fps-example-repo/tree/main/src/mocks). Copy each file in this tutorial into your project, or clone the [fps-example-repo](https://github.com/solidity-labs-io/fps-example-repo/) repo. This tutorial assumes you have cloned the fps-example-repo.

Reviewer: ```suggestion
After adding FPS into project dependencies, the next step is the creation of the first Proposal contract. This example provides guidance on writing a proposal for deploying new instances of `Vault.sol` and `Token`. These contracts are located in the fps-example-repo [mocks](https://github.com/solidity-labs-io/fps-example-repo/tree/main/src/mocks). Clone the [fps-example-repo](https://github.com/solidity-labs-io/fps-example-repo/) repo before continuing the tutorial.
```

Author: Done

        if (!addresses.isAddressSet("MULTISIG_VAULT")) {
            Vault multisigVault = new Vault();

            addresses.addAddress(
                "MULTISIG_VAULT",
                address(multisigVault),
                true
            );
            multisigVault.transferOwnership(multisig);
        }

        // Deploy token address if not already deployed, transfer ownership to multisig
        // and transfer all initial minted tokens from deployer to multisig.
        if (!addresses.isAddressSet("MULTISIG_TOKEN")) {
            Token token = new Token();
            addresses.addAddress("MULTISIG_TOKEN", address(token), true);
            token.transferOwnership(multisig);

            // During forge script execution, the deployer of the contracts is
            // the DEPLOYER_EOA. However, when running through forge test, the deployer of the contracts is this contract.
            uint256 balance = token.balanceOf(address(this)) > 0
                ? token.balanceOf(address(this))
                : token.balanceOf(addresses.getAddress("DEPLOYER_EOA"));

            token.transfer(multisig, balance);
        }
    }

>     ```

Reviewer: Since these changes do not persist from runs themselves, after the contracts are deployed, the user must update the Addresses.json file with the newly deployed contract addresses.

Author: Done

> -   `build()`: Set the necessary actions for your proposal, [Refer](../overview/architecture/proposal-functions.md#build-function). In this example, ERC20 token is whitelisted on the Vault contract. Then multisig approves token for vault and deposits all tokens into the vault. The actions should be written in solidity code and in the order they should be executed. Any calls (except to the Addresses object) will be recorded and stored as actions to execute in the run function. `caller` address that will call actions is passed into `buildModifier`, it is the multisig for this example. `buildModifier` is necessary modifier for `build` function and will not work without it.

Reviewer: refer to

The `caller`

Author: done

        address token = addresses.getAddress("MULTISIG_TOKEN");

        // Get multisig's token balance
        uint256 balance = Token(token).balanceOf(address(multisig));

        /// CALLS -- mutative and recorded

        // Whitelists the deployed token on the deployed vault.
        Vault(multisigVault).whitelistToken(token, true);

        // Approve the token for the vault.
        Token(token).approve(multisigVault, balance);

        // Deposit all tokens into the vault.
        Vault(multisigVault).deposit(token, balance);
    }
    ```

-   `simulate()`: Execute the proposal actions outlined in the `build()` step. This function performs a call to `simulateActions()` from the inherited `MultisigProposal` contract. Internally, `_simulateActions()` simulates a call to the [Multicall3](https://www.multicall3.com/) contract with the calldata generated from the actions set up in the build step.

    ```solidity
    function simulate() public override {
        // Get multisig address
        address multisig = addresses.getAddress("DEV_MULTISIG");

        // multisig is the caller for all the proposal actions
        _simulateActions(multisig);
    }
    ```

> -   `validate()`: This final step validates the system in its post-execution state. It ensures that the multisig is the new owner of Vault and token, the tokens were transferred to multisig, and the token was whitelisted on the Vault contract

Reviewer: ```suggestion
-   `validate()`: This final step validates the system in its post-execution state. It ensures that the multisig is the new owner of Vault and token, the tokens were transferred to the multisig, and the token was whitelisted on the Vault contract
```

        // Get token address
        Token token = Token(addresses.getAddress("MULTISIG_TOKEN"));

        // Get multisig address
        address multisig = addresses.getAddress("DEV_MULTISIG");

        // Ensure token total supply is 10 million
        assertEq(token.totalSupply(), 10_000_000e18);

        // Ensure multisig is owner of deployed token.
        assertEq(token.owner(), multisig);

        // Ensure multisig is owner of deployed vault
        assertEq(multisigVault.owner(), multisig);

        // Ensure vault is not paused
        assertFalse(multisigVault.paused());

        // Ensure token is whitelisted on vault
        assertTrue(multisigVault.tokenWhitelist(address(token)));

        // Get vault's token balance
        uint256 balance = token.balanceOf(address(multisigVault));

        // Get multisig deposits in vault
        (uint256 amount, ) = multisigVault.deposits(address(token), multisig);

        // Ensure multisig deposit is same as vault's token balance
        assertEq(amount, balance);

        // Ensure all minted tokens are deposited into the vault
        assertEq(token.balanceOf(address(multisigVault)), token.totalSupply());
    }
    ```

> -   `run()`: Sets environment for running the proposal. [Refer](../overview/architecture/proposal-functions.md#run-function). It sets `addresses`, `primaryForkId` and calls `super.run()` to run proposal lifecycle. In this function, `primaryForkId` is set to `sepolia` and selecting the fork for running proposal. Next `addresses` object is set by reading `addresses.json` file. `addresses` contract state is persisted accross forks using `vm.makePersistent()`.

Reviewer: refer to

The `addresses`

Author: Done

    ```solidity
    function build()
        public
        override
        buildModifier(addresses.getAddress("DEV_MULTISIG"))
    {
        /// STATICCALL -- non-mutative and hence not recorded for the run stage

        // Get multisig address
        address multisig = addresses.getAddress("DEV_MULTISIG");

        // Get vault address
        address multisigVault = addresses.getAddress("MULTISIG_VAULT");

        // Get token address
        address token = addresses.getAddress("MULTISIG_TOKEN");

        // Get multisig's token balance
        uint256 balance = Token(token).balanceOf(address(multisig));

        /// CALLS -- mutative and recorded

        // Whitelists the deployed token on the deployed vault.
        Vault(multisigVault).whitelistToken(token, true);

        // Approve the token for the vault.
        Token(token).approve(multisigVault, balance);

        // Deposit all tokens into the vault.
        Vault(multisigVault).deposit(token, balance);
    }
    ```

> -   `simulate()`: Execute the proposal actions outlined in the `build()` step. This function performs a call to `simulateActions()` from the inherited `MultisigProposal` contract. Internally, `_simulateActions()` simulates a call to the [Multicall3](https://www.multicall3.com/) contract with the calldata generated from the actions set up in the build step.

Reviewer: In each proposal document, you should make clear the difference between `run()` and `simulate()` as it's a bit confusing, and if it takes me a second to wrap my head around it, users that are completely new to this will take a lot longer.

This should be explained across msig, timelock, and governor bravo contracts.

My understanding is that the simulate step is for tests, and the run function is for actually broadcasting the deployment.

Reviewer 2: run is not necessary for broadcasting as it can be used to generate calldata as well. simulate can be called inside run and by test files

        assertEq(token.balanceOf(address(multisigVault)), token.totalSupply());
    }
    ```

-   `run()`: Sets environment for running the proposal. It sets `addresses` and `primaryForkId`. `addresses` is address object containing addresses to be used in proposal that are fetched from `Addresses.json`. `primaryForkId` is the RPC URL or alias of the blockchain that will be used to simulate the proposal actions and broadcast if any contract deployment is required.

    ```solidity
    function run() public override {
        // Create sepolia fork for proposal execution. This fork will be selected while running proposal in super.run().
        primaryForkId = vm.createFork("sepolia");

        // Set addresses object reading addresses from json file.
        setAddresses(
            new Addresses(
                vm.envOr("ADDRESSES_PATH", string("addresses/Addresses.json"))
            )
        );

        // Make 'addresses' state persist across selected fork.
        vm.makePersistent(address(addresses));

        // Call the run function of parent contract 'Proposal.sol'.
        super.run();
    }
    ```

## Proposal simulation

> First of all, remove all addresses in `Addresses.json` before running through the tutorial. To proceed with execution, there are two options available:

Reviewer: ```suggestion
First, remove all addresses in `Addresses.json` before running through the tutorial. There are two ways to execute proposals:
```

Reviewer: let's apply this suggestion to all guides

Author: done

## Overview

> Following the addition of FPS to project dependencies, the next step is creating a Proposal contract. This example serves as a guide for drafting a proposal for Multisig contract.

Reviewer: ```suggestion
Following the addition of FPS to project dependencies, the next step is creating a Proposal contract. This example serves as a guide for drafting a proposal for a Multisig contract.
```

> Following the addition of FPS to project dependencies, the subsequent step involves creating the initial Proposal contract. This example serves as a guide for drafting a proposal for Multisig contract.

Reviewer: ```suggestion
Following the addition of FPS to project dependencies, the next step is creating a Proposal contract. This example serves as a guide for drafting a proposal for Multisig contract.
```

                address(multisigVault),
                true
            );
            multisigVault.transferOwnership(multisig);
        }

        // Deploy token address if not already deployed, transfer ownership to multisig
        // and transfer all initial minted tokens from deployer to multisig.
        if (!addresses.isAddressSet("MULTISIG_TOKEN")) {
            Token token = new Token();
            addresses.addAddress("MULTISIG_TOKEN", address(token), true);
            token.transferOwnership(multisig);

            // During forge script execution, the deployer of the contracts is
            // the DEPLOYER_EOA. However, when running through forge test, the deployer of the contracts is this contract.
            uint256 balance = token.balanceOf(address(this)) > 0
                ? token.balanceOf(address(this))
                : token.balanceOf(addresses.getAddress("DEPLOYER_EOA"));

            token.transfer(multisig, balance);
        }
    }
    ```

    Since these changes do not persist from runs themselves, after the contracts are deployed, the user must update the Addresses.json file with the newly deployed contract addresses.

> -   `build()`: Add actions to the proposal contract. [See build function](../overview/architecture/proposal-functions.md#build-function). In this example, an ERC20 token is whitelisted on the Vault contract. Then multisig approves token for vault and deposits all tokens into the vault. The actions should be written in solidity code and in the order they should be executed. Any calls (except to the Addresses object) will be recorded and stored as actions to execute in the run function. The `caller` address that will call actions is passed into `buildModifier`, it is the multisig for this example. `buildModifier` is necessary modifier for `build` function and will not work without it.

Reviewer: ```suggestion
-   `build()`: Add actions to the proposal contract. [See the build function](../overview/architecture/proposal-functions.md#build-function). In this example, an ERC20 token is whitelisted on the Vault contract. Then the multisig approves the token to be spent by the vault, and calls deposit on the vault. The actions should be written in solidity code and in the order they should be executed in the proposal. Any calls (except to the Addresses and Foundry Vm contract) will be recorded and stored as actions to execute in the run function. The `caller` address that will call actions is passed into `buildModifier`, it is the multisig for this example. The `buildModifier` is a necessary modifier for the `build` function, and will not work without it.
```

Reviewer: this same feedback should be applied to all proposals

> -   `build()`: Add actions to the proposal contract. [See the build function](../overview/architecture/proposal-functions.md#build-function). In this example, an ERC20 token is whitelisted on the Vault contract. Then the multisig approves the token to be spent by the vault, and calls deposit on the vault. The actions should be written in solidity code and in the order they should be executed in the proposal. Any calls (except to the Addresses and Foundry Vm contract) will be recorded and stored as actions to execute in the run function. The `caller` address that will call actions is passed into `buildModifier`, it is the multisig for this example. The `buildModifier` is necessary modifier for `build` function and will not work without it.

Reviewer: link should be at the end like previous comments

    ```solidity
    function build()
        public
        override
        buildModifier(addresses.getAddress("DEV_MULTISIG"))
    {
        /// STATICCALL -- non-mutative and hence not recorded for the run stage

        // Get multisig address
        address multisig = addresses.getAddress("DEV_MULTISIG");

        // Get vault address
        address multisigVault = addresses.getAddress("MULTISIG_VAULT");

        // Get token address
        address token = addresses.getAddress("MULTISIG_TOKEN");

        // Get multisig's token balance
        uint256 balance = Token(token).balanceOf(address(multisig));

        /// CALLS -- mutative and recorded

        // Whitelists the deployed token on the deployed vault.
        Vault(multisigVault).whitelistToken(token, true);

        // Approve the token for the vault.
        Token(token).approve(multisigVault, balance);

        // Deposit all tokens into the vault.
        Vault(multisigVault).deposit(token, balance);
    }
    ```

> -   `run()`: Sets up the environment for running the proposal. [See the run function](../overview/architecture/proposal-functions.md#run-function). This sets `addresses`, `primaryForkId` and calls `super.run()` run the entire proposal. In this example, `primaryForkId` is set to `sepolia` and selecting the fork for running proposal. Next the `addresses` object is set by reading from the `Addresses.json` file.

Reviewer: link should be at the end like previous comments

    ```solidity
    function run() public override {
        // Create and select sepolia fork for proposal execution
        primaryForkId = vm.createFork("sepolia");
        vm.selectFork(primaryForkId);

        // Set addresses object reading addresses from json file.
        setAddresses(
            new Addresses(
                vm.envOr("ADDRESSES_PATH", string("addresses/Addresses.json"))
            )
        );

        // Call the run function of parent contract 'Proposal.sol'.
        super.run();
    }
    ```

> -   `simulate()`: Execute the proposal actions outlined in the `build()` step. This function performs a call to `_simulateActions()` from the inherited `MultisigProposal` contract. Internally, `_simulateActions()` simulates a call to the [Multicall3](https://www.multicall3.com/) contract with the calldata generated from the actions set up in the build step.

Reviewer: I feel like this could be more clearly explained that multicall3 is used to batch multiple actions to gether through a single safe action with a delegatecall

        assertEq(token.owner(), multisig);

        // Ensure multisig is owner of deployed vault
        assertEq(multisigVault.owner(), multisig);

        // Ensure vault is not paused
        assertFalse(multisigVault.paused());

        // Ensure token is whitelisted on vault
        assertTrue(multisigVault.tokenWhitelist(address(token)));

        // Get vault's token balance
        uint256 balance = token.balanceOf(address(multisigVault));

        // Get multisig deposits in vault
        (uint256 amount, ) = multisigVault.deposits(address(token), multisig);

        // Ensure multisig deposit is same as vault's token balance
        assertEq(amount, balance);

        // Ensure all minted tokens are deposited into the vault
        assertEq(token.balanceOf(address(multisigVault)), token.totalSupply());
    }
    ```

## Proposal simulation

### Deploying a Gnosis Safe Multisig on Testnet

To kick off this tutorial, a Gnosis Safe Multisig contract is needed to be set up on the testnet.

1. Go to [Gnosis Safe](https://app.safe.global/) and pick your preferred testnet (Sepolia is used for this tutorial). Follow the on-screen instructions to generate a new Safe Account.

> 2. After setting up of Safe, address can be found in the details section of the Safe Account. Make sure to copy this address and keep it handy for later steps.

Reviewer: ```suggestion
2. After setting up the Safe, its address can be found in the details section of the Safe Account. Make sure to copy this address and keep it handy for later steps.
```

-   `name()`: Define the name of your proposal.
-   `description()`: Provide a detailed description of your proposal.
-   `deploy()`: Deploy any necessary contracts. This example demonstrates the
    deployment of Vault and an ERC20 token. Once the contracts are deployed,
    they are added to the `Addresses` contract by calling `addAddress()`.

> -   `run()`: Sets environment for running the proposal. It sets `addresses` and `primaryForkId`. `addresses` is address object

Reviewer: same comment as governor bravo proposal

Author: added code snippet with inline comments and also added reference to `proposal-functions.md` file.

    containing addresses to be used in proposal that are fetched from `Addresses.json`. `primaryForkId` is the RPC URL or alias of the blockchain that will be used to simulate the proposal actions and broadcast if any contract deployment is required.

> -   `build()`: Set the necessary actions for your proposal. In this example,

Reviewer: build should come before run

Author: I think `run` should come either at the beginning or at the end. In the beginning as the proposal environment is setup in run and in the end because it runs all the other functions. Should this be moved to the end or right after build. @ElliotFriedman @anajuliabit

Reviewer: My bias is to talk about the system exactly as it is.

If we describe the steps in order, I think it will make much more sense to end users. If we talk about things out of order, then people might get confused because things do not logically flow from on thing to the next.

Reviewer: there are many places where this is still not addressed

Author: Done

    ERC20 token is whitelisted on the Vault contract. The actions should be
    written in solidity code and in the order they should be executed. Any calls (except to the Addresses object) will be recorded and stored as actions to execute in the run function. `caller` address that will call actions is passed into `buildModifier`, it is the multisig for this example.
-   `simulate()`: Execute the proposal actions outlined in the `build()` step. This
    function performs a call to `simulateActions()` from the inherited
    `MultisigProposal` contract. Internally, `_simulateActions()` simulates a call to the [Multicall3](https://www.multicall3.com/) contract with the calldata generated from the actions set up in the build step.
-   `validate()`: This final step is crucial for validating the post-execution state. It ensures that the multisig is the new owner of Vault and token, the tokens were transferred to multisig and the token was whitelisted on the Vault contract

## Proposal simulation

> First of all, please remove all addresses in `Addresses.json` before running through the tutorial. Now it's time to proceed with execution. There are two options available:

Reviewer: same comments as gov bravo proposal

Author: done
